spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return undefined.
+        1. Let |entries| be the result of running [=map/getting the values=] on |databaseMap|.
+        1. Return |entries|.
+  </div>
+
+  <div algorithm>
+    To <dfn for="shared storage database">count entries in the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
+
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return undefined.
+        1. Let |count| be |databaseMap|'s [=map/size=].
+        1. Return |count|.
+  </div>
 
-[Exposed=(Window)]
-interface SharedStorage {
-  Promise<any> set(DOMString key, DOMString value, optional SharedStorageSetMethodOptions options = {});
+  <div algorithm>
+    To <dfn for="shared storage database">purge expired entries from the database</dfn>, given a [=/browsing context=] |context|, run the following steps:
 
-  Promise<any> append(DOMString key, DOMString value);
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return false.
+        1. For each [=entry/key=] |key| in |databaseMap|:
+            1. Let |valueStruct| be the result of running [=map/Get=] on |databaseMap| with |key|.
+            1. Let |isExpired| be the result of running [=shared storage database/determine whether an entry is expired=] with |valueStruct|.
+            1. If |isExpired| is true, [=map/Remove=] |databaseMap|[|key|].
+        1. Return true.
+  </div>
 
-  Promise<any> delete(DOMString key);
+  <div algorithm>
+    To <dfn for="shared storage database">determine whether an entry is expired</dfn>, given a [=entry/value struct=] |valueStruct|, run the following steps:
 
-  Promise<any> clear();
+    1. Let |lastUpdated| be |valueStruct|'s [=value struct/last updated=].
+    1. Let |lifetime| be [=user agent=]'s [=default entry lifetime=].
+    1. Let |expiration| be the sum of |lastUpdated| and |lifetime|.
+    1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+    1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
```

Comment:
I think you can use `[=current wall time=]` instead of these two steps.  It gets the unsafe current time and automatically runs it through the coarsen time step.